crates/pallet-domains/src/runtime_registry.rs

```diff
@@ -73,10 +93,18 @@ pub(crate) fn can_upgrade_code(
 pub(crate) fn do_register_runtime<T: Config>(
     runtime_name: Vec<u8>,
     runtime_type: RuntimeType,
-    code: Vec<u8>,
+    raw_genesis_storage: Vec<u8>,
     at: BlockNumberFor<T>,
 ) -> Result<RuntimeId, Error> {
-    let runtime_version = runtime_version(&code)?;
+    let mut raw_genesis: RawGenesis = Decode::decode(&mut raw_genesis_storage.as_slice())
+        .map_err(|_| Error::FailedToDecodeRawGenesis)?;
+
+    // Remove the runtime code from `raw_genesis` before storing it
```

Comment:
Because runtime code is updatable, it is weird to store updated runtime code in the genesis state.

Reply:
> Because runtime code is updatable, it is weird to extract the code, update it, and store the updated code back to the genesis state.

This should never be necessary, runtime code in genesis state doesn't need to be updatable. It needs to be canonical instead and come as a bundle, just like it is before you remove runtime code out of it. For updated runtime genesis state will be different too. State and code should come in pairs when it comes to instantiation.

You don't update genesis state without updating runtime code and you don't update runtime code without updating genesis state.

Reply:
Okay, I may be misled by your reply https://github.com/subspace/subspace/pull/1824#discussion_r1320247718, that is what I thought, the genesis state and the runtime code should come in pairs, this should be true for both the first registration and later upgrade. The new domain instance should use both the upgraded runtime code and genesis state to initialize its genesis block, this is the only place the genesis state is involved. The existing domain instance should just upgrade its runtime code don't care about the updated genesis state.

Reply:
The last part of above comment is correct. They come together, but during upgrades we only upgrade code and let it change the state accordingly.

Reply:
The existing domain instance will definitely upgrade its code and change its state accordingly, what I mean is for domain instances that are instantiated after the runtime code is upgraded in the consensus chain, they should use both the upgraded code and genesis state to initialize their genesis block.

Reply:
Hm, it depends on how it is implemented. The way I see it is when code changes it is essentially a different version of domain, so you upload both new genesis state and code and then point existing domain to the new code. What you're suggesting seems to be that domain code is updated "in place", which is confusing for other domains that use the old code (will old code be just "gone" or swapped forcefully with a new version?), but also implies that any future version of the code should be able to upgrade from any past version of the state, which at some point might take so long it wouldn't instantiate at all (potentially).

But keep in mind that I don't have latest design of domains in my head.

Reply:
I think it makes sense to leave the Initial runtime in the Genesis state itself. Sure it helps in retrieving the domain runtime faster. But apart from that, I dont see any benefit but rather brings confusion where is a new runtime. 
I say, we store all the new runtime seperately as they are submitted but leave the first one in place. This also avoids the manually refilling them when we need the full genesis

Reply:
It still means we need to update state one step at a time when we deploy new domain. Maybe this is acceptable, but feels inefficient.

Reply:
I'm okay with storing the runtime code along with the genesis state, but what we need to discuss is whether the genesis state should be updated as the domain runtime upgraded in the consensus chain. My opinion is yes, the domain genesis state and runtime code should come in pairs and be updated at the same time.

> What you're suggesting seems to be that domain code is updated "in place", which is confusing for other domains that use the old code
> I say, we store all the new runtime seperately as they are submitted but leave the first one in place. This also avoids the manually refilling them when we need the full genesis

Suppose the domain is instantiated, it commits to a runtime object with the specified `RuntimeId`, and uses both the runtime object's runtime code and genesis state to initialize the domain chain.

Later, the runtime object is updated (both the code and genesis state), and the existing domain instance will upgrade its runtime code unconditionally and automatically as the upgrade is derived from the consensus block deterministically. However, the existing domain doesn't care about the updated genesis state because it only needs this when building its genesis block.

Then there is a new domain instance instantiated and committed to the same runtime object with the same `RuntimeId`, now the new domain instance should use the newest version of both the runtime code and genesis state to initialize its genesis block. This is the reason why I think we should update the runtime object's runtime code and genesis state at the same time.

And I can't see why we need to store all the version of the runtime code in the consensus chain state to support instantiating new domain instances with an older version of the runtime code.

Reply:
> My opinion is yes, the domain genesis state and runtime code should come in pairs and be updated at the same time.

That is my expectation and preference as well.

> And I can't see why we need to store all the version of the runtime code in the consensus chain state to support instantiating new domain instances with an older version of the runtime code.

Then whatever the latest is will have to support runtime upgrades from whatever the oldest registered domain is (even if it is not running right now) because otherwise domain will not resume. I feel like this is an issue and storing all *used* versions is very desirable.

Reply:
> the oldest registered domain is (even if it is not running right now)

No, as said above the domain runtime upgrade is not controlled by the domain owner but by the runtime object, whenever the runtime object is upgraded the domain instance will upgrade its runtime unconditionally and automatically.

And the domain instance node is not running for a while, then it should catch up to the domain chain tip by deriving the domain block from the consensus block, it can't miss any consensus block since it is instantiated unless we support syncing the domain chain from other nodes, but still all the consensus block should be handled by some (but not all) domain nodes in case any potential bundle/runtime upgrade is missed.

Reply:
The PR is updated with runtime code and genesis state stored together and updated together, PTAL.